src/ipvs/ip_vs_redirect.c

```diff
@@ -0,0 +1,409 @@
+
+static struct rte_ring    *dp_vs_redirect_ring[DPVS_MAX_LCORE][DPVS_MAX_LCORE];
+
+#ifdef CONFIG_DPVS_IPVS_DEBUG
+static inline void
+dp_vs_redirect_show(struct dp_vs_redirect *r, const char *action)
+{
+    char sbuf[64], dbuf[64];
+
+    if (r->proto != IPPROTO_ICMPV6) {
+        RTE_LOG(DEBUG, IPVS, "[%d] redirect %s: [%d] %s %s/%d -> %s/%d\n",
+                rte_lcore_id(), action, r->cid,
+                inet_proto_name(r->proto),
+                inet_ntop(r->af, &r->saddr, sbuf, sizeof(sbuf)) ? sbuf : "::",
+                ntohs(r->sport),
+                inet_ntop(r->af, &r->daddr, dbuf, sizeof(dbuf)) ? dbuf : "::",
+                ntohs(r->dport));
+    }
+}
+#endif
+
+struct dp_vs_redirect *
+dp_vs_redirect_alloc(enum dpvs_fwd_mode fwdmode)
+{
+    struct dp_vs_redirect *r;
+
+    if (dp_vs_redirect_disable) {
+        return NULL;
+    }
+
+    /*
+     * Currently, IPv6 support has the below issues.
+     * a) Fdir IPv6 rules fail to be created with "perfect" mode, but can be
+     * created with "signature" mode.
+     *
+     * b) In full-nat mode, the packets from incoming direction and outgoing
+     *   direction are dispatched to the different cores so the service is
+     *   broken.
+     *
+     * The solutuion is to use decentralized packet dispatch for the symemtric
+     * service modes, full-nat/snat/nat before issue a) is fixed.
+     */
+    if (fwdmode != DPVS_FWD_MODE_FNAT
+        && fwdmode != DPVS_FWD_MODE_SNAT
+        && fwdmode != DPVS_FWD_MODE_NAT) {
+        return NULL;
+    }
+
+    if (unlikely(rte_mempool_get(this_cr_cache, (void **)&r) != 0)) {
+        RTE_LOG(WARNING, IPVS,
+                "%s: no memory for redirect\n", __func__);
+        return NULL;
+    }
+
+    memset(r, 0, sizeof(struct dp_vs_redirect));
+    r->redirect_pool = this_cr_cache;
+
+    return r;
+}
+
+void dp_vs_redirect_free(struct dp_vs_conn *conn)
+{
+    if (conn->redirect) {
+#ifdef CONFIG_DPVS_IPVS_DEBUG
+        dp_vs_redirect_show(conn->redirect, "free");
+#endif
+        rte_mempool_put(this_cr_cache, conn->redirect);
+        conn->redirect = NULL;
+    }
+}
+
+void dp_vs_redirect_hash(struct dp_vs_conn *conn)
+{
+    uint32_t hash;
+    struct dp_vs_redirect *r = conn->redirect;
+
+    if (!r || unlikely(dp_vs_conn_is_redirect_hashed(conn))) {
+        return;
+    }
+
+    hash = dp_vs_conn_hashkey(conn->af,
+                    &tuplehash_out(conn).saddr, tuplehash_out(conn).sport,
+                    &tuplehash_out(conn).daddr, tuplehash_out(conn).dport,
+                    DPVS_CR_TBL_MASK);
+
+    rte_spinlock_lock(&dp_vs_cr_lock[hash]);
+    list_add(&r->list, &dp_vs_cr_tbl[hash]);
+    rte_spinlock_unlock(&dp_vs_cr_lock[hash]);
+
+    dp_vs_conn_set_redirect_hashed(conn);
+}
+
+void dp_vs_redirect_unhash(struct dp_vs_conn *conn)
+{
+    uint32_t hash;
+    struct dp_vs_redirect *r = conn->redirect;
+
+    if (r && likely(dp_vs_conn_is_redirect_hashed(conn))) {
+        hash = dp_vs_conn_hashkey(r->af,
```

Comment:
In NAT64 of NAT/FullNAT，`r->af` in `dp_vs_redirect_unhash` is AF_INET(refer to `dp_vs_redirect_init`), while `conn->af` is AF_INET6 in `dp_vs_redirect_hash`(refer to `dp_vs_conn_fill_param` and `dp_vs_conn_new`). The hashkey do not match.

Reply:
Only one single directional redirect is installed in the global redirect table. The reason is that each core searches its local connection hash table for the packet, if not found, then search the global redirect table. So there is no need to install bi-directional redirects in the global redirect table.

- In fnat/nat NAT64 case, only the outgoing redirect is installed.
- In IPv6-to-IPv6 SNAT case, only the incoming redirect is installed.